lib/encryptionmanager.cpp

```diff
@@ -0,0 +1,206 @@
+            : olmAccount(new Account(encryptionAccountPickle)), // TODO: passphrase even with qtkeychain?
+              signedKeysProportion(move(signedKeysProportion)),
+              oneTimeKeyThreshold(move(oneTimeKeyThreshold)),
+              targetKeysNumber(olmAccount->maxOneTimeKeys()) // 2 // see note below
+        {
+            Q_ASSERT((0 <= signedKeysProportion) && (signedKeysProportion <= 1));
+            Q_ASSERT((0 <= oneTimeKeyThreshold) && (oneTimeKeyThreshold <= 1));
+            /*
+             * Note about targetKeysNumber:
+             *
+             * From: https://github.com/Zil0/matrix-python-sdk/
+             * File: matrix_client/crypto/olm_device.py
+             *
+             * Try to maintain half the number of one-time keys libolm can hold uploaded
+             * on the HS. This is because some keys will be claimed by peers but not
+             * used instantly, and we want them to stay in libolm, until the limit is reached
+             * and it starts discarding keys, starting by the oldest.
+             */
+        }
+        ~Private()
+        {
+            delete olmAccount;
+        }
+
+        UploadKeysJob* uploadIdentityKeysJob = nullptr;
+        UploadKeysJob* uploadOneTimeKeysJob = nullptr;
+
+        Account* olmAccount;
+        const QByteArray encryptionAccountPickle;
+
+        float signedKeysProportion;
+        float oneTimeKeyThreshold;
+        int targetKeysNumber;
+
+        void updateKeysToUpload();
+        bool oneTimeKeyShouldUpload();
+
+        QHash<QString, int> oneTimeKeyCounts;
+        void setOneTimeKeyCounts(const QHash<QString, int> oneTimeKeyCountsNewValue)
+        {
+            oneTimeKeyCounts = oneTimeKeyCountsNewValue;
+            updateKeysToUpload();
+        }
+        QHash<QString, int> oneTimeKeysToUploadCounts;
+        QHash<QString, int> targetOneTimeKeyCounts
+        {
+            {SignedCurve25519Name, qRound(signedKeysProportion * targetKeysNumber)},
+            {Curve25519Name, qRound((1-signedKeysProportion) * targetKeysNumber)}
+        };
+};
+
+EncryptionManager::EncryptionManager(const QByteArray &encryptionAccountPickle, float signedKeysProportion, float oneTimeKeyThreshold,
+                                     QObject* parent)
+    : QObject(parent),
+    d(std::make_unique<Private>(std::move(encryptionAccountPickle), std::move(signedKeysProportion), std::move(oneTimeKeyThreshold)))
+{
+
+}
+
+EncryptionManager::~EncryptionManager() = default;
+
+void EncryptionManager::uploadIdentityKeys(Connection* connection)
+{
+    // https://matrix.org/docs/spec/client_server/latest#post-matrix-client-r0-keys-upload
+    DeviceKeys deviceKeys
+    {
+        /*
+         * The ID of the user the device belongs to. Must match the user ID used when logging in.
+         * The ID of the device these keys belong to. Must match the device ID used when logging in.
+         * The encryption algorithms supported by this device.
+         */
```

Comment:
Well, what I really expected is just a one-liner that this block is `keys` and the next one is `signatures`, for people without IDEs that give parameter hints. But ok, let's be verbose here.